mangaki/mangaki/models.py

```diff
@@ -165,38 +165,62 @@ def get_absolute_url(self):
+        anidb_tags = anidb.handle_tags(anidb_aid=self.anidb_aid)
 
-        tag_work = TaggedWork.objects.filter(work=self)
-        current_tags = {tagwork.tag.title: tagwork.weight for tagwork in tag_work}
+        tag_work_list = TaggedWork.objects.filter(work=self).all()
+        values = tag_work_list.values_list('tag__title', 'tag__anidb_tag_id', 'weight')
+        current_tags = {value[0]: {"weight": value[2], "anidb_tag_id": value[1]} for value in values}
 
         deleted_tags_keys = current_tags.keys() - anidb_tags.keys()
         deleted_tags = {key: current_tags[key] for key in deleted_tags_keys}
 
         added_tags_keys = anidb_tags.keys() - current_tags.keys()
         added_tags = {key: anidb_tags[key] for key in added_tags_keys}
 
-        remaining_tags_keys = anidb_tags.keys() & current_tags.keys()
+        remaining_tags_keys = list(set(current_tags.keys()).intersection(anidb_tags.keys()))
         remaining_tags = {key: current_tags[key] for key in remaining_tags_keys}
 
         updated_tags = {title: (current_tags[title], anidb_tags[title]) for title in remaining_tags if current_tags[title] != anidb_tags[title]}
         kept_tags = {title: current_tags[title] for title in remaining_tags if current_tags[title] == anidb_tags[title]}
 
         return {"deleted_tags": deleted_tags, "added_tags": added_tags, "updated_tags": updated_tags, "kept_tags": kept_tags}
 
-    def update_tags(self, deleted_tags, added_tags, updated_tags):
-        for title, weight in added_tags.items():
-            current_tag = Tag.objects.get_or_create(title=title)[0]
-            TaggedWork(tag=current_tag, work=self, weight=weight).save()
+    def update_tags(self, anidb_tags):
+        existing_tags = TaggedWork.objects.filter(work=self).all()
+        values = existing_tags.values_list('tag__title', 'tag__anidb_tag_id', 'weight')
+        existing_tags = {value[0]: {"weight": value[2], "anidb_tag_id": value[1]} for value in values}
 
-        tags = Tag.objects.filter(title__in=updated_tags.keys())
-        for tag in tags:
-            tagged_work = self.taggedwork_set.get(tag=tag)
-            tagged_work.weight = updated_tags[tag.title][1]
+        new_tags_keys = anidb_tags.keys() - existing_tags.keys()
+        new_tags = {key: anidb_tags[key] for key in new_tags_keys}
+
+        old_tags_keys = list(set(anidb_tags.keys()).intersection(existing_tags.keys()))
+        old_tags = {key: existing_tags[key] for key in old_tags_keys}
+        old_tags_updated = {key: anidb_tags[key] for key in old_tags_keys}
+
+        deleted_tags_keys = existing_tags.keys() - anidb_tags.keys()
+
+        # New tags have to be added to the database (if they aren't already present)
+        # New tags have to be assigned to that work
+        for title, tag_infos in new_tags.items():
+            tag, created = Tag.objects.get_or_create(title=title, anidb_tag_id=tag_infos["anidb_tag_id"])
```

Comment:
Rather than creating the tag then the tagged work, as we know those are new tags, we should bulk_create them beforehand (the tags), then bulk_create the tagged work.
This way, it reduces dramatically an amount of queries.

Plan:

1. Prepare the data for all tags in an array, `bulk_create` them.
2. `bulk_create` the `TaggedWork`.